sql-to-dbsp-compiler/SQL-compiler/src/main/codegen/includes/ddl.ftl

```diff
@@ -0,0 +1,184 @@
+    final Span s = Span.of();
+    final ColumnStrategy strategy;
+}
+{
+    LOOKAHEAD(2) id = SimpleIdentifier()
+    (
+        type = DataType()
+        nullable = NullableOptDefaultTrue()
+        (
+            <DEFAULT_> e = Expression(ExprContext.ACCEPT_SUB_QUERY) {
+                strategy = ColumnStrategy.DEFAULT;
+            }
+        |
+            {
+                e = null;
+                strategy = nullable ? ColumnStrategy.NULLABLE
+                    : ColumnStrategy.NOT_NULLABLE;
+            }
+        )
+        {
+            list.add(
+                SqlDdlNodes.column(s.add(id).end(this), id,
+                    type.withNullable(nullable), e, strategy));
+        }
+    |
+        { list.add(id); }
+    )
+|
+    id = SimpleIdentifier() {
+        list.add(id);
+    }
+|
+    [ <CONSTRAINT> { s.add(this); } name = SimpleIdentifier() ]
+    (
+        <CHECK> { s.add(this); } <LPAREN>
+        e = Expression(ExprContext.ACCEPT_SUB_QUERY) <RPAREN> {
+            list.add(SqlDdlNodes.check(s.end(this), name, e));
+        }
+    |
+        <UNIQUE> { s.add(this); }
+        columnList = ParenthesizedSimpleIdentifierList() {
+            list.add(SqlDdlNodes.unique(s.end(columnList), name, columnList));
+        }
+    |
+        <PRIMARY>  { s.add(this); } <KEY>
+        columnList = ParenthesizedSimpleIdentifierList() {
+            list.add(SqlDdlNodes.primary(s.end(columnList), name, columnList));
+        }
+    )
+}
+
+SqlNodeList AttributeDefList() :
+{
+    final Span s;
+    final List<SqlNode> list = new ArrayList<SqlNode>();
+}
+{
+    <LPAREN> { s = span(); }
+    AttributeDef(list)
+    (
+        <COMMA> AttributeDef(list)
+    )*
+    <RPAREN> {
+        return new SqlNodeList(list, s.end(this));
+    }
+}
+
+void AttributeDef(List<SqlNode> list) :
+{
+    final SqlIdentifier id;
+    final SqlDataTypeSpec type;
+    final boolean nullable;
+    SqlNode e = null;
+    final Span s = Span.of();
+}
+{
+    id = SimpleIdentifier()
+    (
+        type = DataType()
+        nullable = NullableOptDefaultTrue()
+    )
+    [ <DEFAULT_> e = Expression(ExprContext.ACCEPT_SUB_QUERY) ]
+    {
+        list.add(SqlDdlNodes.attribute(s.add(id).end(this), id,
+            type.withNullable(nullable), e, null));
+    }
+}
+
+SqlCreate SqlCreateType(Span s, boolean replace) :
+{
+    final SqlIdentifier id;
+    SqlNodeList attributeDefList = null;
+    SqlDataTypeSpec type = null;
+}
+{
+    <TYPE>
+    id = CompoundIdentifier()
+    <AS>
+    (
+        attributeDefList = AttributeDefList()
+    |
+        type = DataType()
+    )
+    {
+        return SqlDdlNodes.createType(s.end(this), replace, id, attributeDefList, type);
+    }
+}
+
+SqlCreate SqlCreateTable(Span s, boolean replace) :
+{
+    final boolean ifNotExists;
+    final SqlIdentifier id;
+    SqlNodeList tableElementList = null;
+    SqlNode query = null;
+}
+{
+    <TABLE> ifNotExists = IfNotExistsOpt() id = CompoundIdentifier()
```

Comment:
I am not sure that "if exists" makes sense in our context
where is the CREATE keyword?

Reply:
If people are copy pasting DDL from their databases, it's fine to let it in?